Support all date buckets on waterfall charts

alxnddr opened the pull request:
Closes https://github.com/metabase/metabase/issues/37320
Closes https://github.com/metabase/metabase/issues/37305

### Description

This PR adds timeseries unit inferrence from data even for native queries which lets us compute the waterfall total x-axis value correctly. In addition to that, it lets us group timeline events by unit on native queries.

### How to verify

- Create a timeseries waterfall chart in GUI
- Test it on all datetime units and ensure it looks good
- Create a timeseries waterfall chart using an SQL query
- Ensure it correctly infers bucket from the data and bar widths are sensible

Timeline events
Ensure timeline events on native queries are grouped correctly. For instance if data contains monthly granularity and you have two events Jan 1st and Jan 31st, they should be grouped under the tick that represents January.

Sample query (converted from GUI)
```
SELECT
  DATE_TRUNC('month', "PUBLIC"."ORDERS"."CREATED_AT") AS "CREATED_AT",
  COUNT(*) AS "count",
  AVG("PUBLIC"."ORDERS"."TOTAL") AS "avg"
FROM
  "PUBLIC"."ORDERS"
WHERE
  (
    "PUBLIC"."ORDERS"."CREATED_AT" >= timestamp with time zone '2023-04-01 00:00:00.000-03:00'
  )
 
   AND (
    "PUBLIC"."ORDERS"."CREATED_AT" < timestamp with time zone '2023-10-01 00:00:00.000-03:00'
  )
GROUP BY
  DATE_TRUNC('month', "PUBLIC"."ORDERS"."CREATED_AT")
ORDER BY
  DATE_TRUNC('month', "PUBLIC"."ORDERS"."CREATED_AT") ASC
```
Add timeline events on 1 Aug 2023 and 31 Aug 2023 and ensure they are grouped (or check the screenshots below).

### Demo
Github does not upload my images, so I uploaded images to a random image uploading service.

Before
https://pasteboard.co/0aVN3tYHQhzW.png

After
https://pasteboard.co/UvKhrRkhGfFq.png

**Timeline events**

Before
https://pasteboard.co/ppI1XHvRCIgl.png

After
https://pasteboard.co/UGYlvfDgLpJb.png

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

alxnddr commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/metabase/metabase/37908?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `master`
  * **PR #36489** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36489?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37908** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37908?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>  👈
    * **PR #37903** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37903?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37901** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37901?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37871** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37871?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37847** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37847?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37843** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37843?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37842** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37842?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37835** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37835?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37776** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37776?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37773** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37773?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37772** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37772?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37728** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37728?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37721** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37721?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #37743** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37743?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37680** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37680?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37673** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37673?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37621** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37621?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #37769** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37769?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37603** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37603?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37543** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37543?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37538** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37538?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37489** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37489?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37488** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37488?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37487** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37487?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37486** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37486?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37282** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37282?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37277** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37277?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37276** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37276?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37247** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37247?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37196** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37196?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37143** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37143?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37142** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37142?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37094** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37094?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37088** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37088?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37064** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37064?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37063** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37063?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36961** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36961?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36957** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36957?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36926** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36926?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36925** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36925?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36888** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36888?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36887** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36887?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36886** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36886?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36884** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36884?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36883** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36883?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36882** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36882?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36879** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36879?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36878** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36878?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36781** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36781?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36779** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36779?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36728** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36728?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36621** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36621?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36564** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36564?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36560** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36560?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36548** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36548?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36510** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36510?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36502** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36502?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36486** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36486?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36317** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36317?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36135** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36135?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #36563** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36563?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
        * **PR #36729** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36729?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36013** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36013?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35997** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35997?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35923** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35923?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35922** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35922?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35783** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35783?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #35996** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35996?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35573** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35573?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35395** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35395?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35345** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35345?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35321** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35321?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35307** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35307?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35304** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35304?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35303** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35303?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35301** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35301?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35300** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35300?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35297** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35297?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35118** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35118?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #34820** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34820?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #34923** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34923?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #34717** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34717?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
        * **PR #34772** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34772?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
          * **PR #34809** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34809?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
            * **PR #34816** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34816?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
              * **PR #34859** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34859?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
                * **PR #34869** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34869?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
                  * **PR #34872** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34872?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
                    * **PR #35068** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35068?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #34703** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34703?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
        * **PR #34763** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34763?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/metabase/metabase/37908?utm_source=stack-comment).

alxnddr committed:
waterfall supports all time series units

alxnddr committed:
timeline events bucketing my unit for native questions

alxnddr committed:
snapshots

alxnddr committed:
x-axis key

EmmadUsmani commented:
https://github.com/metabase/metabase/assets/37751258/552886eb-8847-4854-876a-e7f297a8771f

Seeing some weird behavior on certain buckets for this question

Week: total label is on the wrong value

Quarter: some labels are misaligned and there are duplicate labels

Year of era: question errors

EmmadUsmani commented on frontend/src/metabase/visualizations/echarts/cartesian/model/axis.ts:
```diff
@@ -461,13 +472,77 @@ export function getXAxisModel(
+  settings: ComputedVisualizationSettings,
+) {
+  if (
+    ["linear", "log", "pow", "histogram"].includes(
```
Shouldn't this be the opposite condition? If it's non numeric (e.g. timeseries or ordinal), then we don't compute the numeric interval? The dc.js code does it this way

```javascript
  } else if (isQuantitative(settings) || isHistogram(settings)) {
    // Get the bin width from binning_info, if available
    // TODO: multiseries?
    const binningInfo =
      getFirstNonEmptySeries(series).data.cols[0].binning_info;
    if (binningInfo) {
      return binningInfo.bin_width;
    }

    // Otherwise try to infer from the X values
    return computeNumericDataInverval(xValues);
  }
```

EmmadUsmani commented on frontend/src/metabase/visualizations/echarts/cartesian/model/axis.ts:
```diff
@@ -461,13 +472,77 @@ export function getXAxisModel(
+) {
+  if (
+    ["linear", "log", "pow", "histogram"].includes(
+      // @ts-expect-error x axis scale value can be undefined
```
We can just include an `undefined` check rather than expecting an error

EmmadUsmani commented on frontend/src/metabase/visualizations/echarts/cartesian/model/axis.ts:
```diff
@@ -461,13 +472,77 @@ export function getXAxisModel(
+  return { count, interval, timezone };
+}
+
+export function getNumericXAxisInterval(
```
Let's remove this function from this PR, I don't see `numericInterval` being used anywhere.

EmmadUsmani commented on frontend/src/metabase/visualizations/echarts/cartesian/model/types.ts:
```diff
@@ -69,9 +78,17 @@ export type Extent = [number, number];
 export type XAxisModel = {
   label?: string;
   formatter: AxisFormatter;
+  numericInterval?: number;
```
As mentioned before, let's remove this for now since we aren't using it

EmmadUsmani commented on frontend/src/metabase/visualizations/echarts/cartesian/waterfall/model/dataset.ts:
```diff
@@ -98,8 +122,16 @@ export function getWaterfallDataset(
+      xAxisModel,
+    );
+
+    if (totalTimeSeriesXValue == null) {
```
It would be better to handle errors closer to the source within `getTotalTimeSeriesXValue` instead of separating it here, e.g.

```typescript
const getTotalTimeSeriesXValue = (
  lastDimensionValue: RowValue,
  { timeSeriesInterval }: XAxisModel,
) => {
  if (typeof lastDimensionValue === "boolean") {
    throw Error(
      "dimension value cannot be boolean with timeseries x-axis scale",
    );
  }
  if (timeSeriesInterval == null) {
    throw Error("Missing timeSeriesInterval for timeseries waterfall chart");
  }
```

Also, I think the types are incorrect. `XAxisModel.timeSeriesInterval` will never be undefined right now, since we provide a default in `getTimeSeriesXAxisInterval`:

```typescript
  const { count, interval } = (computeTimeseriesDataInverval(xValues, unit) ?? {
    count: 1,
    interval: "day",
  }) as Pick<TimeSeriesInterval, "count" | "interval">;
  return { count, interval, timezone };
```

We should either
A) make the type non-optional if we want to provide a default
B) don't provide a default and keep the type optional

alxnddr replied:
`XAxisModel.timeSeriesInterval` has become optional in the next PR: https://github.com/metabase/metabase/pull/37947/files#diff-a925af260455ef56ff9bc97d2376ab8cc1f92bc51256d5ab0fdb4b39a9088363R506

alxnddr replied:
Updated `getTotalTimeSeriesXValue` here  https://github.com/metabase/metabase/pull/37947/commits/3f80c752cb99309c8b60555bfb26c93f8b72fe25

alxnddr commented:
@EmmadUsmani thanks for the feedback, I addressed it. Dropped `numericInterval` for now.
Regarding "Year of era", I don't see this bucket. What DB do you use? Does it work on other charts?

alxnddr commented:
> Week: total label is on the wrong value

I spent a bit of time trying to fix it with known workarounds but it seems ECharts is not letting us control it so with week bucketing it looks bad on all charts. I'll create a separate issue to track this.
<img width="1363" alt="Screenshot 2024-02-15 at 9 45 18 PM" src="https://github.com/metabase/metabase/assets/14301985/356d6cc0-577b-4b04-937f-5aee8fee86c3">
